tidy3d/plugins/mode/mode_solver.py

```diff
@@ -116,6 +116,58 @@ def solve(self) -> ModeSolverData:
         """
         return self.data
 
+    def solve_group_index(self, frac_step: pydantic.PositiveFloat = 0.005) -> ModeIndexDataArray:
+        r"""Calculate the group index.
+
+        The group index is numerically calculated as
+        $$ n_g = n_\text{eff} + f \frac{\mathrm d n_\text{eff}}{\mathrm df} $$
+        with a central difference algorithm used to approximate the differential.
+
+        Parameters
+        ----------
+        frac_step : float
+            Fractional frequency step used for numerical differentiation.
+
+        Returns
+        -------
+        :class:`ModeIndexDataArray`
+            Group indices for all waveguide wavelengths and modes indices.
+        """
+        if frac_step >= 1:
+            raise ValidationError("Argument `frac_step` must be less than 1.")
+
+        if self.mode_spec.track_freq is None:
+            log.warning(
+                "Group index calculation without mode tracking can lead to incorrect results "
+                "around mode crossings."
+            )
+
+        if self.mode_spec.precision != "double":
+            log.warning(
+                "Group index calculation should be performed with double precision for better "
+                "accuracy."
+            )
+
+        # We have to solve all frequencies in a single waveguide because we want to take advantage
+        # of mode tracking.
+        freqs = np.outer(self.freqs, (1 - frac_step, 1, 1 + frac_step)).flatten()
+        end = freqs.size
+
+        mode_solver = self.copy(update={"freqs": freqs})
+        n_backward = mode_solver.data.n_eff.isel(f=slice(0, end, 3)).values
+        n_center = mode_solver.data.n_eff.isel(f=slice(1, end, 3)).values
+        n_forward = mode_solver.data.n_eff.isel(f=slice(2, end, 3)).values
+
+        n_group = n_center + (n_forward - n_backward) / (2 * frac_step)
```

Comment:
I guess if we want to save 1/3 of the solves we could also do

```
n_center = (n_forward + n_backward) / 2
n_group = n_center + (n_forward - n_backward) / (2 * frac_step)
```

I would think this should not have any noticeable effect on precision, what do you think?

Reply:
Honestly I'm not a big fan of this… It's hard to say how much of an impact it has (each waveguide and wavelength region would be impacted differently) but there's definitely some impact. I think a 3-point derivative is not that expensive and gives reasonable results (I actually though about using a 5-point derivative, but given the errors in the n_eff, I thought it might be too expensive for no real gain).
Now, I know I'm kind of conservative with these things and tend to err on the side of caution, so my opinion might not be what most users want…

Reply:
Fair enough, also it may be moot depending on what we decide about whether to have it as part of the regular `solve`, or as a standalone `solve_group_index`.